### PR TITLE
Move GeoAxes.background_patch to GeoAxes.patch.

### LINKED ISSUE #1207: tight_layout and Matplotlib 3
### Description

Recently, I was working on a multi panel map figure. The figure used `tight_layout`. The figures produced in Matplotlib 2 have smaller margins than those in Matplotlib 3. 

As @tacaswell [pointed out on Twitter](https://twitter.com/tacaswell/status/1057633159141982211) in a thread with @dopplershift , “More things now participate it [sic] tight_layout.” A warning message received while running suggests that perhaps additional Cartopy “axes decorations” participate in tight_layout in Matplotlib 3.

Below, I include four plots (both MPL versions; with and without Caropy axes) that demonstrate the problem. The upstream changes in Matplotlib 3 apparently result in more than cosmetic layout differences on Cartopy axes for these relatively simple plots.

2 2 3_no_map
3 0 2_no_map

#### Code to reproduce

I created two minimal environments:

```bash
conda create -n mpl2 -c conda-forge matplotlib=2 python=3.7 cartopy

conda create -n mpl3 -c conda-forge matplotlib=3 python=3.7 cartopy
```

Then using the plotting script, I received the following output.

```bash
ebruning$ source activate mpl3
(mpl3) ebruning$ python plot.py 
Plotting with map
/Users/ebruning/anaconda/envs/mpl3/lib/python3.7/site-packages/matplotlib/tight_layout.py:176: UserWarning: Tight layout not applied. The left and right margins cannot be made large enough to accommodate all axes decorations. 
  warnings.warn('Tight layout not applied. The left and right margins '
Plotting with no map
(mpl3) ebruning$ source deactivate
ebruning$ source activate mpl2
(mpl2) ebruning$ python plot.py 
Plotting with map
Plotting with no map
```

The plotting script
```python
import numpy as np
import cartopy.crs as ccrs

import matplotlib
matplotlib.use('Agg')
ver = matplotlib.__version__

import matplotlib.pyplot as plt

### This code produces smaller margins in Matplotlib 2 vs. 3.
print("Plotting with map")
fig = plt.figure()

subplots = (2,3)
n_panels = subplots[0] * subplots[1]

lons = np.arange(-120, -80, 1, dtype=float)
lats = np.arange(20, 50, 1, dtype=float)
datashape = lons.shape+lats.shape

fields = [np.random.randn(*datashape) for i in range(n_panels)]
imgextent=(lons.min(), lons.max(), lats.min(), lats.max())

proj = ccrs.PlateCarree()
for fi, f in enumerate(fields):
    ax = fig.add_subplot(subplots[0], subplots[1], fi+1, projection=proj)
    ax.imshow(f, extent=imgextent)
    ax.coastlines()
fig.suptitle('Map, matplotlib={0}'.format(ver))
fig.tight_layout()
fig.savefig(ver+'.png')

### Demonstrate that the issue only happens with Caropy axes.
### This code produces identical figures in Matplotlib 2 and 3.
print("Plotting with no map")
fignomap = plt.figure()
subplots = (2,3)
n_panels = subplots[0] * subplots[1]

lons = np.arange(-120, -80, 1, dtype=float)
lats = np.arange(20, 50, 1, dtype=float)
datashape = lons.shape+lats.shape

fields = [np.random.randn(*datashape) for i in range(n_panels)]
imgextent=(lons.min(), lons.max(), lats.min(), lats.max())
for fi, f in enumerate(fields):
    ax = fignomap.add_subplot(subplots[0], subplots[1], fi+1)
    ax.imshow(f, extent=imgextent)
fignomap.suptitle('No map, matplotlib={0}'.format(ver))
fignomap.tight_layout()
fignomap.savefig(ver+'_no_map.png')
```

#### Traceback 

No errors.

<details>
 <summary>Full environment definition</summary>

### Operating system

Mac OS X 10.14 (Mojave). Behavior also observed on CentOS server.

### Cartopy version

0.17.0

### conda list

*Matplotlib 2*
```
(mpl2) ebruning$ conda list
# packages in environment at /Users/ebruning/anaconda/envs/mpl2:
#
asn1crypto                0.24.0                py37_1003    conda-forge
blas                      1.0                         mkl  
ca-certificates           2018.10.15           ha4d7672_0    conda-forge
cartopy                   0.17.0           py37h81b52dc_0    conda-forge
certifi                   2018.10.15            py37_1000    conda-forge
cffi                      1.11.5           py37h5e8e0c9_1    conda-forge
chardet                   3.0.4                 py37_1003    conda-forge
cryptography              2.3.1            py37hdffb7b8_0    conda-forge
cycler                    0.10.0                     py_1    conda-forge
freetype                  2.9.1                h6debe1e_4    conda-forge
geos                      3.6.2                hfc679d8_3    conda-forge
icu                       58.2                 hfc679d8_0    conda-forge
idna                      2.7                   py37_1002    conda-forge
intel-openmp              2019.0                      118  
jpeg                      9c                   h470a237_1    conda-forge
kiwisolver                1.0.1            py37h2d50403_2    conda-forge
libcxx                    7.0.0                h2d50403_2    conda-forge
libffi                    3.2.1                hfc679d8_5    conda-forge
libgfortran               3.0.1                h93005f0_2  
libiconv                  1.15                 h470a237_3    conda-forge
libpng                    1.6.35               ha92aebf_2    conda-forge
libtiff                   4.0.9                he6b73bb_2    conda-forge
libxml2                   2.9.8                h422b904_5    conda-forge
libxslt                   1.1.32               h88dbc4e_2    conda-forge
llvm-meta                 7.0.0                         0    conda-forge
lxml                      4.2.5            py37hc9114bc_0    conda-forge
matplotlib                2.2.3            py37h0e0179f_0    conda-forge
mkl                       2019.0                      118  
mkl_fft                   1.0.6            py37hb8a8100_0  
mkl_random                1.0.1            py37h5d10147_1  
ncurses                   6.1                  hfc679d8_1    conda-forge
numpy                     1.15.4           py37h6a91979_0  
numpy-base                1.15.4           py37h8a80b8c_0  
olefile                   0.46                       py_0    conda-forge
openssl                   1.0.2p               h470a237_1    conda-forge
owslib                    0.17.0                     py_0    conda-forge
pillow                    5.3.0            py37hc736899_0    conda-forge
pip                       18.1                  py37_1000    conda-forge
proj4                     4.9.3                h470a237_8    conda-forge
pycparser                 2.19                       py_0    conda-forge
pyepsg                    0.3.2                      py_1    conda-forge
pykdtree                  1.3.1            py37h7eb728f_2    conda-forge
pyopenssl                 18.0.0                py37_1000    conda-forge
pyparsing                 2.3.0                      py_0    conda-forge
pyproj                    1.9.5.1          py37h508ed2a_6    conda-forge
pyshp                     2.0.0                      py_0    conda-forge
pysocks                   1.6.8                 py37_1002    conda-forge
python                    3.7.1                h46c1a51_0    conda-forge
python-dateutil           2.7.5                      py_0    conda-forge
pytz                      2018.7                     py_0    conda-forge
readline                  7.0                  haf1bffa_1    conda-forge
requests                  2.20.1                py37_1000    conda-forge
scipy                     1.1.0            py37h28f7352_1  
setuptools                40.6.2                   py37_0    conda-forge
shapely                   1.6.4            py37h164cb2d_0    conda-forge
six                       1.11.0                py37_1001    conda-forge
sqlite                    3.25.3               hb1c47c0_0    conda-forge
tk                        8.6.9                ha92aebf_0    conda-forge
tornado                   5.1.1            py37h470a237_0    conda-forge
urllib3                   1.23                  py37_1001    conda-forge
wheel                     0.32.2                   py37_0    conda-forge
xz                        5.2.4                h470a237_1    conda-forge
zlib                      1.2.11               h470a237_3    conda-forge
```

*Matplotlib 3*
```
(mpl3) ebruning$ conda list
# packages in environment at /Users/ebruning/anaconda/envs/mpl3:
#
asn1crypto                0.24.0                py37_1003    conda-forge
blas                      1.0                         mkl  
ca-certificates           2018.10.15           ha4d7672_0    conda-forge
cartopy                   0.17.0           py37h81b52dc_0    conda-forge
certifi                   2018.10.15            py37_1000    conda-forge
cffi                      1.11.5           py37h5e8e0c9_1    conda-forge
chardet                   3.0.4                 py37_1003    conda-forge
cryptography              2.3.1            py37hdffb7b8_0    conda-forge
cycler                    0.10.0                     py_1    conda-forge
freetype                  2.9.1                h6debe1e_4    conda-forge
geos                      3.6.2                hfc679d8_3    conda-forge
icu                       58.2                 hfc679d8_0    conda-forge
idna                      2.7                   py37_1002    conda-forge
intel-openmp              2019.0                      118  
jpeg                      9c                   h470a237_1    conda-forge
kiwisolver                1.0.1            py37h2d50403_2    conda-forge
libcxx                    7.0.0                h2d50403_2    conda-forge
libffi                    3.2.1                hfc679d8_5    conda-forge
libgfortran               3.0.1                h93005f0_2  
libiconv                  1.15                 h470a237_3    conda-forge
libpng                    1.6.35               ha92aebf_2    conda-forge
libtiff                   4.0.9                he6b73bb_2    conda-forge
libxml2                   2.9.8                h422b904_5    conda-forge
libxslt                   1.1.32               h88dbc4e_2    conda-forge
llvm-meta                 7.0.0                         0    conda-forge
lxml                      4.2.5            py37hc9114bc_0    conda-forge
matplotlib                3.0.2                    py37_1    conda-forge
matplotlib-base           3.0.2            py37hb2d221d_1    conda-forge
mkl                       2019.0                      118  
mkl_fft                   1.0.6            py37hb8a8100_0  
mkl_random                1.0.1            py37h5d10147_1  
ncurses                   6.1                  hfc679d8_1    conda-forge
numpy                     1.15.4           py37h6a91979_0  
numpy-base                1.15.4           py37h8a80b8c_0  
olefile                   0.46                       py_0    conda-forge
openssl                   1.0.2p               h470a237_1    conda-forge
owslib                    0.17.0                     py_0    conda-forge
pillow                    5.3.0            py37hc736899_0    conda-forge
pip                       18.1                  py37_1000    conda-forge
proj4                     4.9.3                h470a237_8    conda-forge
pycparser                 2.19                       py_0    conda-forge
pyepsg                    0.3.2                      py_1    conda-forge
pykdtree                  1.3.1            py37h7eb728f_2    conda-forge
pyopenssl                 18.0.0                py37_1000    conda-forge
pyparsing                 2.3.0                      py_0    conda-forge
pyproj                    1.9.5.1          py37h508ed2a_6    conda-forge
pyshp                     2.0.0                      py_0    conda-forge
pysocks                   1.6.8                 py37_1002    conda-forge
python                    3.7.1                h46c1a51_0    conda-forge
python-dateutil           2.7.5                      py_0    conda-forge
pytz                      2018.7                     py_0    conda-forge
readline                  7.0                  haf1bffa_1    conda-forge
requests                  2.20.1                py37_1000    conda-forge
scipy                     1.1.0            py37h28f7352_1  
setuptools                40.6.2                   py37_0    conda-forge
shapely                   1.6.4            py37h164cb2d_0    conda-forge
six                       1.11.0                py37_1001    conda-forge
sqlite                    3.25.3               hb1c47c0_0    conda-forge
tk                        8.6.9                ha92aebf_0    conda-forge
tornado                   5.1.1            py37h470a237_0    conda-forge
urllib3                   1.23                  py37_1001    conda-forge
wheel                     0.32.2                   py37_0    conda-forge
xz                        5.2.4                h470a237_1    conda-forge
zlib                      1.2.11               h470a237_3    conda-forge
```

### pip list

*Matplotlib 2*
```
(mpl2) ebruning$ pip list
Package         Version    
--------------- -----------
asn1crypto      0.24.0     
Cartopy         0.17.0     
certifi         2018.10.15 
cffi            1.11.5     
chardet         3.0.4      
cryptography    2.3.1      
cycler          0.10.0     
idna            2.7        
kiwisolver      1.0.1      
lxml            4.2.5      
matplotlib      2.2.3      
mkl-fft         1.0.6      
mkl-random      1.0.1      
numpy           1.15.4     
olefile         0.46       
OWSLib          0.17.0     
Pillow          5.3.0      
pip             18.1       
pycparser       2.19       
pyepsg          0.3.2      
pykdtree        1.3.1      
pyOpenSSL       18.0.0     
pyparsing       2.3.0      
pyproj          1.9.5.1    
pyshp           2.0.0      
PySocks         1.6.8      
python-dateutil 2.7.5      
pytz            2018.7     
requests        2.20.1     
scipy           1.1.0      
setuptools      40.6.2     
Shapely         1.6.4.post2
six             1.11.0     
tornado         5.1.1      
urllib3         1.23       
wheel           0.32.2     
```

*Matplotlib 3*
```
(mpl3) ebruning$ pip list
Package         Version    
--------------- -----------
asn1crypto      0.24.0     
Cartopy         0.17.0     
certifi         2018.10.15 
cffi            1.11.5     
chardet         3.0.4      
cryptography    2.3.1      
cycler          0.10.0     
idna            2.7        
kiwisolver      1.0.1      
lxml            4.2.5      
matplotlib      3.0.2      
mkl-fft         1.0.6      
mkl-random      1.0.1      
numpy           1.15.4     
olefile         0.46       
OWSLib          0.17.0     
Pillow          5.3.0      
pip             18.1       
pycparser       2.19       
pyepsg          0.3.2      
pykdtree        1.3.1      
pyOpenSSL       18.0.0     
pyparsing       2.3.0      
pyproj          1.9.5.1    
pyshp           2.0.0      
PySocks         1.6.8      
python-dateutil 2.7.5      
pytz            2018.7     
requests        2.20.1     
scipy           1.1.0      
setuptools      40.6.2     
Shapely         1.6.4.post2
six             1.11.0     
tornado         5.1.1      
urllib3         1.23       
wheel           0.32.2     
```
</details>

### PR DESCRIPTION
**Pull request by QuLogic**
## Rationale

This is a followup to #1213, converting the Cartopy-specific `background_patch` into the usual `Axes.patch` from Matplotlib.

## Implications

Unfortunately, this breaks several tests. Most of these are due to clipping bugs upstream in Matplotlib (at least matplotlib/matplotlib#15952). I'm hoping to get these fixed, regenerate any test images with the new version if necessary, then just increase tolerance for the old images (_most_ of them are small changes anyway.)

Since we aren't using the default patch, this should fix tight layout too, but I haven't yet tested it.

Fixes #1207.

### CODE REVIEW

**Comment by dopplershift** [comment]
In general this looks good. Everything seems cleaner. Waiting to see what the image changes look like..

**Comment by QuLogic** [comment]
Currently, it's mostly the same as the referenced Matplotlib issue, plus matplotlib/matplotlib#15946.

**Comment by greglucas** [comment]
Looks great to me. I just pulled your code and this does appear to fix the tight layout issues. Specifically, #1207 looks good to me locally with your changes.

With constrained_layout it appears much better, but there is still a slight shift between the two figures.

```
import cartopy
import cartopy.crs as ccrs
import matplotlib.pyplot as plt

f, ax1 = plt.subplots(1, 1, constrained_layout=True,
                      subplot_kw=dict(projection=ccrs.PlateCarree()))

f.savefig('mpl-1.png')
f.savefig('mpl-2.png')
```

**Comment by QuLogic** [comment]
For reference, I've committed the changed images and the diffs [here](https://github.com/QuLogic/cartopy/commit/67da84c0841b49f190983b7bf3d4bd173b88dc18). To understand most of the changes, you have to know that Matplotlib draws spines at 0.8 linewidth, so there's usually one middle pixel full black, and two surrounding pixels of grey. Correctly clipped artists should end at the inner grey or the black pixel, and the inner grey one might be shaded a bit by the artist's colour. If not clipped correctly, then they might mix into the outer grey pixels, or only mix into the inner grey one via antialiasing (and thus a bit darker than they should be.)

My analysis of the changes is:
* `contour_label.png` 0.9719825101307121. The difference is along the left/right edges, but honestly can't see it.
* `gshhs_coastlines.png` 0.006666666666666667. Most tests use 0.5 tolerance, so I just increased that. It seems like some of the coastlines near the edge changed, so I'm not exactly sure why, but it could be because of the way it clipped.
* `regrid_image.png` 5.5476486310267825. Mainly the clipping now works for the first Axes. However, there is an annoying extra bit of clipped pixels on the red/green/blue images, mostly visible on the left edges.
* `global_map.png` 0.5430546319969413. The line here extends outside the spines just a little bit. It should be clipped, but it's visible on the outer grey side.
* `multiple_projections5.png` 0.5250993556778882. This is mostly just stuff around the edges, but there are a lot of subplots so it raises the RMS. For the other lines, I suspect that things were already different, but RMS was under the original 0.5 tolerance before.
* `pcolormesh_global_wrap1.png` 1.2638301573655641. The pcolormesh extends outside the Axes on the left and left/top edges.
* `pcolormesh_global_wrap2.png` 1.6054726285773642. Same as above.
* `pcolormesh_global_wrap3.png` 1.3892948851045908. Same as above, though I think the Robinson map is a bit better instead.
* `pcolormesh_goode_wrap.png` 0.6293873785056563. Basically the same edge changes. I think this one is also improved a bit.
* `nightshade_platecarree.png ` 1.1415358805866185. The inner grey edge along the top is much darker now and it doesn't match the inner grey on the left and right edges.
* `poly_interiors.png` 3.0048004417006537. This is outright a bug in Matplotlib. https://github.com/matplotlib/matplotlib/issues/15946

**Comment by greglucas** [comment]
Are you going to close this draft PR and merge all the changes in that other link you sent?
The changed images all look good to me. Minor and likely due to edge antialiasing as you mentioned. The only question I have is in regards to your last bullet.

> poly_interiors.png 3.0048004417006537. This is outright a bug in Matplotlib. matplotlib/matplotlib#15946

Do you want to update this image if it is a known bug, or just xfail it in pytest for now and then update the image once that is fixed in MPL upstream?

**Comment by greglucas** [comment]
@QuLogic, are you still planning on doing more work on this? I think it is an excellent improvement and well worth getting in for the next release.

**Comment by QuLogic** [comment]
I still want to do this for 0.18, but I'm looking to fix the two bugs in Matplotlib first. I had half-fixed one, and the other should be possible, though it needs further testing.

**Comment by QuLogic** [comment]
I've now run through the whole combination of dependencies, and updated the tests to account for the change here. I updated the images where stuff was improved and where tolerances were small-ish, I just increased them (note our default is 0.5 RMS.) Images were generated with the latest Matplotlib, and tolerances added for older versions if necessary.

Some much older versions of Matplotlib are completely broken for some tests, but rather than fix that, I just decided to mark the tests as `xfail`. We're going to drop support for old Matplotlib anyway.

Unfortunately, I was not able to fix some of the above bugs in Matplotlib for 3.2.0, so I just upped the tolerances for the tests that were affected here.

**Comment by QuLogic** [comment]
Oh, and one other is `regrid_image.png` is not updated yet. The clipping works better for the first Axes, but the extra clipped pixels on the red/green/blue images is a bug somewhere, and I haven't figured out whether it's Matplotlib or Cartopy yet. We can update that image later.

File: lib/cartopy/mpl/slippy_image_artist.py [hunk 1]
```diff
@@ -40,7 +40,7 @@ class SlippyImageArtist(AxesImage):
     def __init__(self, ax, raster_source, **kwargs):
         self.raster_source = raster_source
         super(SlippyImageArtist, self).__init__(ax, **kwargs)
-        self.set_clip_path(ax.background_patch)
+        self.set_clip_path(ax.patch)
```
Comment (greglucas) [inline lib/cartopy/mpl/slippy_image_artist.py:43]: Now that the background_patch is in the axes, won't this artist be clipped automatically by your new `ax.patch`?

File: lib/cartopy/mpl/geoaxes.py [hunk 2]
```diff
@@ -1416,50 +1431,25 @@ def set_boundary(self, path, transform=None, use_as_clip_path=True):
...
             axes' projection.
-        use_as_clip_path : bool, optional
-            Whether axes.patch should be updated.
-            Updating axes.patch means that any artists
-            subsequently created will inherit clipping
-            from this path, rather than the standard unit
-            square in axes coordinates.
 
         """
+        if use_as_clip_path is not None:
```
[the remark this answers is not in the document]
Comment (QuLogic) [reply lib/cartopy/mpl/geoaxes.py:1436]: I prefer that people use Matplotlib's builtin support for clipping, by setting `clip_on=False` on the artists they don't want to be clipped.

The `regrid_image.png` test image is affected by this, but honestly it looks better this way (note, not updated yet due to reasons above.)

File: lib/cartopy/mpl/slippy_image_artist.py (hunk 1 above)
Comment (QuLogic) [reply lib/cartopy/mpl/slippy_image_artist.py]: You're right, I've remove this.

**Pull request merged**